File: python/lombokalgoritma/compression/huffman.py
# LombokAlgoritma — static Huffman coding (SPEC §13.3)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

import heapq
from dataclasses import dataclass

from ..errors import InvalidInputError


@dataclass
class HuffNode:
    """Huffman tree node: a leaf carries ``symbol``; an internal node has both children."""

    freq: int
    symbol: int | None = None
    left: HuffNode | None = None
    right: HuffNode | None = None


@dataclass
class HuffmanResult:
    """Packed code bits (MSB first, zero-padded), bit length, per-byte codes and the tree."""

    encoded: bytes
    bit_length: int
    codes: dict[int, str]
    tree: HuffNode


def huffman_encode(data: bytes | bytearray) -> HuffmanResult:
    """Deterministic static Huffman code: min-heap keyed ``(freq, id)``, leaves by byte value.

    Left edge = '0', right = '1'; a single-symbol input uses code "0".
    """
    codes: dict[int, str] = {}
    if len(data) == 0:
        return HuffmanResult(b"", 0, codes, HuffNode(0))
    freq = [0] * 256
    for b in data:
        freq[b] += 1
    heap: list[tuple[int, int, HuffNode]] = []
    nid = 0
    for s in range(256):
        if freq[s]:
            heap.append((freq[s], nid, HuffNode(freq[s], s)))
            nid += 1
    heapq.heapify(heap)
    while len(heap) > 1:
        fl, _, left = heapq.heappop(heap)
        fr, _, right = heapq.heappop(heap)
        heapq.heappush(heap, (fl + fr, nid, HuffNode(fl + fr, None, left, right)))
        nid += 1
    tree = heap[0][2]
    stack: list[tuple[HuffNode, str]] = [(tree, "")]
    while stack:
        node, code = stack.pop()
        if node.symbol is not None:
            codes[node.symbol] = code or "0"
            continue
        assert node.left is not None and node.right is not None  # noqa: S101 — internal node
        stack.append((node.right, code + "1"))
        stack.append((node.left, code + "0"))
    bits = "".join(codes[b] for b in data)
    bit_length = len(bits)
    padded = bits + "0" * (-bit_length % 8)
    encoded = int(padded, 2).to_bytes(len(padded) // 8, "big") if padded else b""
    return HuffmanResult(encoded, bit_length, dict(sorted(codes.items())), tree)


def huffman_decode(encoded: bytes | bytearray, bit_length: int, tree: HuffNode) -> bytes:
    """Decode ``bit_length`` bits of ``encoded`` with ``tree``.

    Raises:
        InvalidInputError: ``bit_length`` exceeds the input, or an invalid / truncated code.
    """
    if bit_length > len(encoded) * 8:
        raise InvalidInputError("huffman_decode: bit_length exceeds input")
    if bit_length == 0:
        return b""
    if tree.symbol is not None:
        return bytes((tree.symbol,)) * bit_length
    out = bytearray()
    node = tree
    for i in range(bit_length):
        bit = (encoded[i >> 3] >> (7 - (i & 7))) & 1
        nxt = node.right if bit else node.left
        if nxt is None:
            raise InvalidInputError("huffman_decode: invalid code")
        node = nxt
        if node.symbol is not None:
            out.append(node.symbol)
            node = tree
    if node is not tree:
        raise InvalidInputError("huffman_decode: truncated code")
    return bytes(out)

File: typescript/src/compression/huffman.ts
// LombokAlgoritma — Static Huffman coding
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { InvalidInputError } from '../core/errors.js';
import { MinHeap, tupleLess } from '../core/heap.js';

/** Huffman tree node: a leaf carries `symbol`; an internal node has both children. */
export interface HuffNode {
  symbol?: number;
  freq: number;
  left?: HuffNode;
  right?: HuffNode;
}

/** Result of {@link huffmanEncode}. */
export interface HuffmanResult {
  /** Code bits packed MSB-first; the last byte is zero-padded. */
  encoded: Uint8Array;
  /** Number of meaningful bits in `encoded`. */
  bitLength: number;
  /** Code of every symbol present, as a '0'/'1' string, keyed by byte value. */
  codes: Map<number, string>;
  tree: HuffNode;
}

/**
 * Static Huffman code (SPEC §10.3, normative so codes are identical in every port):
 * leaves get ids 0, 1, … in ascending byte value, internal nodes get the next ids in creation
 * order; a min-heap keyed by (freq, id) pops `left` then `right`, merged as
 * {freq: left + right}. Left edge = '0', right = '1'; a single-symbol input uses code "0".
 * v0.1.x re-sorted by frequency only, so ties depended on Map iteration order.
 */
export function huffmanEncode(data: Uint8Array): HuffmanResult {
  const codes = new Map<number, string>();
  if (data.length === 0)
    return { encoded: new Uint8Array(0), bitLength: 0, codes, tree: { freq: 0 } };
  const freq = new Array<number>(256).fill(0);
  for (const b of data) freq[b] = (freq[b] as number) + 1;
  const heap = new MinHeap<[number, number, HuffNode]>((a, b) =>
    tupleLess([a[0], a[1]], [b[0], b[1]]),
  );
  let id = 0;
  for (let s = 0; s < 256; s++) {
    const f = freq[s] as number;
    if (f > 0) heap.push([f, id++, { symbol: s, freq: f }]);
  }
  while (heap.size > 1) {
    const [fl, , left] = heap.pop() as [number, number, HuffNode];
    const [fr, , right] = heap.pop() as [number, number, HuffNode];
    heap.push([fl + fr, id++, { freq: fl + fr, left, right }]);
  }
  const tree = (heap.pop() as [number, number, HuffNode])[2];
  const walk = (node: HuffNode, code: string): void => {
    if (node.symbol !== undefined) {
      codes.set(node.symbol, code === '' ? '0' : code);
      return;
    }
    walk(node.left as HuffNode, `${code}0`);
    walk(node.right as HuffNode, `${code}1`);
  };
  walk(tree, '');
  let bitLength = 0;
  for (const b of data) bitLength += (codes.get(b) as string).length;
  const encoded = new Uint8Array(Math.ceil(bitLength / 8));
  let pos = 0;
  for (const b of data) {
    for (const bit of codes.get(b) as string) {
      if (bit === '1') encoded[pos >> 3] = (encoded[pos >> 3] as number) | (1 << (7 - (pos & 7)));
      pos++;
    }
  }
  return { encoded, bitLength, codes, tree };
}

/** Decode `bitLength` bits of `encoded` with `tree`. @throws RangeError on an invalid stream */
export function huffmanDecode(encoded: Uint8Array, bitLength: number, tree: HuffNode): Uint8Array {
  if (bitLength > encoded.length * 8)
    throw new InvalidInputError('huffmanDecode: bitLength exceeds input');
  const out: number[] = [];
  if (bitLength === 0) return new Uint8Array(0);
  if (tree.symbol !== undefined) {
    for (let i = 0; i < bitLength; i++) out.push(tree.symbol);
    return new Uint8Array(out);
  }
  let node = tree;
  for (let i = 0; i < bitLength; i++) {
    const bit = ((encoded[i >> 3] as number) >> (7 - (i & 7))) & 1;
    const next = bit === 1 ? node.right : node.left;
    if (next === undefined) throw new InvalidInputError('huffmanDecode: invalid code');
    node = next;
    if (node.symbol !== undefined) {
      out.push(node.symbol);
      node = tree;
    }
  }
  if (node !== tree) throw new InvalidInputError('huffmanDecode: truncated code');
  return new Uint8Array(out);
}
